src/properties/trees.c

```diff
@@ -349,3 +349,197 @@ igraph_error_t igraph_is_tree(const igraph_t *graph, igraph_bool_t *res, igraph_
+        }
+        /* register all its neighbours (except its parent) for future processing */
+        IGRAPH_CHECK(igraph_neighbors(graph, &neis, u, mode));
+        ncount = igraph_vector_int_size(&neis);
+
+        for (i = 0; i < ncount; ++i) {
+            igraph_integer_t v = VECTOR(neis)[i];
+            
+            if(mode==IGRAPH_ALL){
+                if (IGRAPH_LIKELY(!VECTOR(*visited)[v])){
+                    IGRAPH_CHECK(igraph_stack_int_push(&stack, v));
+                }
+                /*to check for loop in undirected graph*/
+                else if(v==u){
+                    *res=0;
+                    break;
+                }
+            }
+            else{
+                IGRAPH_CHECK(igraph_stack_int_push(&stack, v));
+            }
+        }
+    }
+    igraph_vector_int_destroy(&neis);
+    igraph_stack_int_destroy(&stack);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup structural
+ * \function igraph_is_forest
+ * \brief Decides whether the graph is a forest.
+ *
+ * An undirected graph is a forest if it has no cycles.
+ * </para><para>
+ *
+ * In the directed case, a possible additional requirement is that a vertex 
+ * should not have more than one parent. Also, edges in each
+ * tree are oriented away from the root (out-tree or arborescence) or all edges
+ * are oriented towards the root (in-tree or anti-arborescence).
+ * This test can be controlled using the \p mode parameter.
+ * </para><para>
+ *
+ * By convention, the null graph (i.e. the graph with no vertices) is considered to be a forest.
+ *
+ * \param graph The graph object to analyze.
+ * \param res Pointer to a logical variable, the result will be stored
+ *        here.
+ * \param roots A vector in which the root nodes will be stored. When \p mode
+ *        is \c IGRAPH_ALL or the graph is undirected, any 1 vertex from each
+ *        component can be the root. When \p mode is \c IGRAPH_OUT
+ *        or \c IGRAPH_IN, all the vertices with zero in- or out-degree,
+ *        respectively are considered as root nodes.
+ * \param mode For a directed graph this specifies whether to test for an
+ *        out-tree, an in-tree or ignore edge directions. The respective
+ *        possible values are:
+ *        \c IGRAPH_OUT, \c IGRAPH_IN, \c IGRAPH_ALL. This argument is
+ *        ignored for undirected graphs.
+ * \return Error code:
+ *        \c IGRAPH_EINVMODE: invalid mode argument.
+ *
+ * Time complexity: At most O(|V|+|E|), the
+ * number of vertices plus the number of edges in the graph.
+ *
+ */
+igraph_error_t igraph_is_forest(const igraph_t *graph, igraph_bool_t *res, 
+                            igraph_vector_int_t *roots, igraph_neimode_t mode) {
+    igraph_vector_bool_t visited;
+    igraph_integer_t visited_count=0;
+    igraph_integer_t vcount, ecount;
+
+    vcount = igraph_vcount(graph);
+    ecount = igraph_ecount(graph);
+    IGRAPH_CHECK(igraph_vector_int_resize(roots, 0));
```

Comment:
Let us use `igraph_vector_int_clear()` here and drop `IGRAPH_CHECK` (because `igraph_vector_int_clear()` never fails and thus does not return an error code).